### PR TITLE
Store the eqv rule separately on the rule db

### PR DESCRIPTION
**Pull request by enadeau**
This will help solving all the situation where we have trouble
recovering an equivalence rule between two label.

### CODE REVIEW

File: comb_spec_searcher/rule_db/base.py [hunk 1]
```diff
@@ -57,10 +65,10 @@ def add(self, start: int, ends: Iterable[int], rule: AbstractRule) -> None:
...
         if is_equiv:
             self.set_equivalent(start, ends[0])
-        if len(ends) != 1 or is_equiv or not self.are_equivalent(start, ends[0]):
-            # to avoid overwriting an equivalence rule with a non-equivalence
-            # rule, we only save if an equivalence rule, or does not have the
-            # # same start -> ends as some equivalence rule.
+            self.eqv_rule_to_strategy[(start, ends)] = rule.strategy
+            self.rule_to_strategy.pop((start, ends), None)
+            self.rule_to_strategy.pop((ends[0], (start,)), None)
+        else:
```
Comment (christianbean) [inline comb_spec_searcher/rule_db/base.py:71]: Should we only add if we know that start and ends are not equivalent?

File: comb_spec_searcher/rule_db/base.py [hunk 2]
```diff
@@ -246,6 +257,11 @@ def get_specification_rules(
     def _get_specification_rules(
         self, label: int, proof_tree_node: Node
     ) -> SpecificationLabelsAndStrats:
+        # for k, v in self.eqv_rule_to_strategy.items():
+        #     print(k, v)
+        # print("="*80)
+        # for k, v in self.rule_to_strategy.items():
+        #     print(k, v)
```
Comment (christianbean) [inline comb_spec_searcher/rule_db/base.py:264]: delete this?

File: comb_spec_searcher/rule_db/base.py [hunk 3]
```diff
@@ -345,7 +361,15 @@ class RuleDB(RuleDBBase):
     def __init__(self) -> None:
         super().__init__()
         self._rule_to_strategy: Dict[Tuple[int, Tuple[int, ...]], AbstractStrategy] = {}
+        self._eqv_rule_to_strategy: Dict[Tuple[int, Tuple[int, ...]], AbstractStrategy]
+        self._eqv_rule_to_strategy = {}
```
Comment (christianbean) [inline comb_spec_searcher/rule_db/base.py:365]: this can be one line

Comment (enadeau) [reply comb_spec_searcher/rule_db/base.py:365]: Yes but the then black formatting make it ugly :smirk:

File: comb_spec_searcher/rule_db/base.py (hunk 1 above)
Comment (enadeau) [reply comb_spec_searcher/rule_db/base.py:71]: I can do that

File: comb_spec_searcher/rule_db/base.py (hunk 2 above)
Comment (enadeau) [reply comb_spec_searcher/rule_db/base.py:264]: I will remove

**Pull request merged**